src/update_agent/actor.rs

```diff
@@ -73,52 +74,92 @@ impl Message for RefreshTick {
 impl Handler<RefreshTick> for UpdateAgent {
     type Result = ResponseActFuture<Self, Result<(), Error>>;
 
-    fn handle(&mut self, _msg: RefreshTick, ctx: &mut Self::Context) -> Self::Result {
-        let tick_timestamp = chrono::Utc::now();
-        LAST_REFRESH.set(tick_timestamp.timestamp());
-
-        trace!("update agent tick, current state: {:?}", self.state);
-        let prev_state = self.state.clone();
-
-        let state_action = match &self.state {
-            UpdateAgentState::StartState => self.tick_initialize(),
-            UpdateAgentState::Initialized => self.tick_report_steady(),
-            UpdateAgentState::ReportedSteady => self.tick_check_updates(),
-            UpdateAgentState::NoNewUpdate => self.tick_check_updates(),
-            UpdateAgentState::UpdateAvailable((release, _)) => {
-                let update = release.clone();
-                self.tick_stage_update(update)
-            }
-            UpdateAgentState::UpdateStaged((release, _)) => {
-                let update = release.clone();
-                self.tick_finalize_update(update)
-            }
-            UpdateAgentState::UpdateFinalized(release) => {
-                let update = release.clone();
-                self.tick_end(update)
+    fn handle(&mut self, _msg: RefreshTick, _ctx: &mut Self::Context) -> Self::Result {
+        // Acquire RwLock to access state.
+        let lock = Arc::clone(&self.state);
+        // We need a clone of `info` because we need to move it into futures to ensure a
+        // long enough lifetime.
+        let update_agent_info = self.info.clone();
+        let state_action = async move {
+            let mut agent_state_guard = lock.write_owned().await;
+            // Consider `LAST_REFRESH` time to be when lock is acquired.
+            let tick_timestamp = chrono::Utc::now();
+            LAST_REFRESH.set(tick_timestamp.timestamp());
+
+            trace!("update agent tick, current state: {:?}", *agent_state_guard);
+            let prev_state = (*agent_state_guard).clone();
+
+            let modify_state_outcome = match prev_state.clone() {
```

Comment:
The `clone()` here is because the left side of the match arms require a partial move; but we need to borrow `prev_state` later in this function